crates/anymap/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+//! Copyright © 2014–2022 Chris Morgan
+//! https://github.com/chris-morgan/anymap/blob/master/COPYING
+//!
+//! This crate provides a safe and convenient store for one value of each type.
+//!
+//! Your starting point is [`Map`]. It has an example.
+//!
+//! # Cargo features
+//!
+//! This crate has two independent features, each of which provides an implementation providing
+//! types `Map`, `AnyMap`, `OccupiedEntry`, `VacantEntry`, `Entry` and `RawMap`:
+//!
+#![cfg_attr(feature = "std", doc = " - **std** (default, *enabled* in this build):")]
+#![cfg_attr(not(feature = "std"), doc = " - **std** (default, *disabled* in this build):")]
+//!   an implementation using `std::collections::hash_map`, placed in the crate root
+//!   (e.g. `anymap::AnyMap`).
+//!
+#![cfg_attr(feature = "hashbrown", doc = " - **hashbrown** (optional; *enabled* in this build):")]
+#![cfg_attr(
+    not(feature = "hashbrown"),
+    doc = " - **hashbrown** (optional; *disabled* in this build):"
+)]
+//!   an implementation using `alloc` and `hashbrown::hash_map`, placed in a module `hashbrown`
+//!   (e.g. `anymap::hashbrown::AnyMap`).
+
+#![warn(missing_docs, unused_results)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::convert::TryInto;
+use core::hash::Hasher;
+
+pub use crate::any::CloneAny;
+
+mod any;
+
+/// A hasher designed to eke a little more speed out, given `TypeId`’s known characteristics.
+///
+/// Specifically, this is a no-op hasher that expects to be fed a u64’s worth of
+/// randomly-distributed bits. It works well for `TypeId` (eliminating start-up time, so that my
+/// get_missing benchmark is ~30ns rather than ~900ns, and being a good deal faster after that, so
+/// that my insert_and_get_on_260_types benchmark is ~12μs instead of ~21.5μs), but will
+/// panic in debug mode and always emit zeros in release mode for any other sorts of inputs, so
+/// yeah, don’t use it! 😀
+#[derive(Default)]
+pub struct TypeIdHasher {
+    value: u64,
+}
+
+impl Hasher for TypeIdHasher {
+    #[inline]
+    fn write(&mut self, bytes: &[u8]) {
+        // This expects to receive exactly one 64-bit value, and there’s no realistic chance of
+        // that changing, but I don’t want to depend on something that isn’t expressly part of the
+        // contract for safety. But I’m OK with release builds putting everything in one bucket
+        // if it *did* change (and debug builds panicking).
+        debug_assert_eq!(bytes.len(), 8);
+        let _ = bytes.try_into().map(|array| self.value = u64::from_ne_bytes(array));
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.value
+    }
+}
+
+#[cfg(any(feature = "std", feature = "hashbrown"))]
+macro_rules! everything {
```

Comment:
We can simplify all of this a bit more, lets inline the macro call, that is just replace it with its expansion and remove the feature flags, basically enabling them unconditionally, as well

Reply:
There is probably a lot more we can strip out of this after that as well, at which point we might as well just move this in to the `stdx` crate as a standalone module instead of puttin this into a separate crate.

Reply:
Thank you for your comment, OK I will work on them.

Reply:
I have moved the code into stdx crate in this commit. Can you review again please?
[f671b0b](https://github.com/rust-lang/rust-analyzer/pull/15618/commits/f671b0b8644aff682616025af5e1a1780ffab0c3)